lib/graphql_metrics/analyzer.rb

```diff
@@ -0,0 +1,193 @@
+module GraphQLMetrics
+  class Analyzer < GraphQL::Analysis::AST::Analyzer
+    FieldTimingCallback = Struct.new(:static_metrics, :callback, keyword_init: true)
+
+    CONTEXT_NAMESPACE = :graphql_metrics_analysis
+    RUNTIME_METRICS_ENABLED = :runtime_metrics_enabled
+
+    QUERY_START_TIME_KEY = :query_start_time
+    QUERY_START_TIME_MONOTONIC_KEY = :query_start_time_monotonic
+    FIELD_TIMING_CALLBACKS_KEY = :field_timing_callbacks
+    ANALYZER_INSTANCE_KEY = :analyzer_instance
+
+    # NOTE: These constants are used both to match tracing keys from graphql-ruby as well as to store in-line and lazy
+    # field resolution timings in context.
+    GRAPHQL_TRACING_FIELD_KEY = 'execute_field'
+    GRAPHQL_TRACING_LAZY_FIELD_KEY = 'execute_field_lazy'
+
+    class << self
+      def current_time
+        Process.clock_gettime(Process::CLOCK_REALTIME)
+      end
+
+      def current_time_monotonic
+        Process.clock_gettime(Process::CLOCK_MONOTONIC)
+      end
+
+      def before_query(query)
+        return if query.context[:skip_graphql_metrics_analysis]
+        return unless query.valid?
+
+        query.context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+          ns[RUNTIME_METRICS_ENABLED] = true
+          ns[FIELD_TIMING_CALLBACKS_KEY] = []
+          ns[QUERY_START_TIME_KEY] = current_time
+          ns[QUERY_START_TIME_MONOTONIC_KEY] = current_time_monotonic
```

Comment:
We want wall clock time for query / field execution start times, since it's UNIX time (seconds since the UNIX epoch, Jan 1 1970).

We want monotonic clock time for determining durations. Monotonic clocks (at least in Ruby) return the seconds since the machine booted. So I think we need both.

```bash
$ ruby -e "now = Process.clock_gettime(Process::CLOCK_MONOTONIC); puts now; puts \"#{now / 86400} days\""
102546.841438
1.1868847388657409 days
$ uptime
16:23  up 1 day,  4:29, 8 users, load averages: 2.18 1.92 1.72
```

Also see https://github.com/apollographql/apollo-tracing#collected-data